text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,269 @@
+
+# Terminology
+
+I'll may shorten `function` to `fn` sometimes.
+
+- **function pointer**: pointer type with the type syntax `fn(?) -> ?` directly pointing at a function, not the type implementing the `Fn[Once/Mut](?) -> ?` traits.
+- **function item** (or just function): a declared function in code. free-standing or associated to a type.
+- **function group**: many non-specific functions with the same signature (params, return type, etc.)
+- **function trait(s)**: the `Fn[Once/Mut](?) -> ?` traits
+- **function type**: the type behind a function, which also implements the function traits.
+- **fixed type**: directly named type, no generic / `impl Trait`.
+- **describe the function type**: write `fn(..) -> ? name` instead of just `fn name`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As we all know, you can refer to a struct by its name and for example implement a trait
+```rust
+struct Timmy;
+impl Person for Timmy {
+    fn greet() {
+        println!("Hey it's me, Timmy!");
+    }
+}
+```
+When we want to target a specific function for a trait implementation, we somehow need to get to the type behind it. 
+Refering to the hidden type is achieved via the following syntax
+```rust
+fn is_positive(a: i32) -> bool { /* ... */ }
+impl MyTrait for fn(i32) -> bool is_positive {
+    /* ... */
+}
+```
+For function signatures, where every parameter/return-type is a fixed type and can be known just by refering to the function (so no generics or `impl Trait` parameters/return type), we can drop the redundant information:
+```rust
+fn is_positive(a: i32) -> bool { /* ... */ }
+impl MyTrait for fn is_positive {
+    /* ... */
+}
+```
+
+> 💡 NOTE: Even when we need to describe the function type but the return type is `()`, we can (just as for function pointers and function traits) drop the `-> ()` from the type. (This should also be added as a lint).
+
+---
+A function with a more complex signature, like a function that specifies `const`, `unsafe` or `extern "ABI"`, we just ignore that when naming the type:
+```rust
+const fn my_fn(a: i32) -> (i16, i16) { .. }
+impl MyTrait for fn my_fn {}
+// or with explicit declaration
+impl MyTrait for fn(i32) -> (i16, i16) my_fn { .. }
+```
+
+When having an async function, we in theory have a `impl Future<Output = ..>` as a return type, which should force us to explicitly declare the function type like so
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl<F: Future<Output = String>> Requestable for fn(PersonID) -> F request_name {
+    /* ... */
+}
+```
+We can take a shortcut and use the `async` keyword, as long as the `Output` assoc type in the Future is still fixed
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for async fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
```

Comment:
I think it might make sense to explain this relating to [the current reference docs on function items](https://doc.rust-lang.org/reference/types/function-item.html), and how we'd update that given that we now _have_ syntax for talking about the function type.